app/models/concerns/order_workflow.rb

```diff
@@ -71,15 +73,26 @@ def update_address(address_params)
   end
 
   def cancel
-    unless update(cancelled_at: Time.current, cancelled_by: current_user)
-      halt errors.full_messages.join(', ')
+    ActiveRecord::Base.transaction do
+      @refund_payment = RefundService.new(self)
+      refund_successful = @refund_payment.refund[:success]
+      @refund = payment.refunds.build(@refund_payment.build_refund_params_hash) if refund_successful
+      unless refund_successful && @refund.save && update(cancelled_at: Time.current, cancelled_by: current_user)
```

Comment:
Does not looks right, we should refund the payment as a last step. Currently in this code everything can rollback even after the amount has been refunded.